### PR TITLE
Add an option to dependencies called 'method'.

### LINKED ISSUE #711: No way to disable special dependency lookups, such as 'sdl2'
This is hardcoded, and it uses sdl2-config first. But what if I want to use my custom build of sdl2 and not the system's? There should be a way to disable these special dependency classes.

### LINKED ISSUE #1472: Unable to use custom Qt
According to a post on the mailing list:
https://groups.google.com/forum/#!topic/mesonbuild/7u4pYlzQj5A
It should:be possible to use a custom Qt by setting PKG_CONFIG_PATH and PATH

> The common approach is to set up multiple build directories, where each is configured to a specific qt installation. Like this:
> 
> mkdir qt51; cd qt51; PATH=/path/to/qt51/bin:$PATH meson .. <args>; cd ..
>mkdir qt52; cd qt52; PATH=/path/to/qt52/bin:$PATH meson .. <args>; cd ..
>mkdir qt53; cd qt53; PATH=/path/to/qt53/bin:$PATH meson .. <args>; cd ..
> 
> Then you can test any installation simply by running Ninja in the corresponding build dir.
> 
> This is a bit of work but you only need to set it up once per project. If that does not work, then please file a bug.

However I have Qt installed on my system, which provides .pc files, and a custom Qt that I want to use for this build which does not.

When the build is configured, it first tries pkg-config, and detects the system Qt module for all of the modules needed in the build that have a system version as well. I can't override pkg-config since I don't have any other .pc files to use. If I could, the fallback to qmake works correctly, but I can't get to that fallback.

### PR DESCRIPTION
**Pull request by absmall**
This can be used to configure a detection method, for those types of dependencies that have
more than one means of detection.

The default detection methods are unchanged if 'method' is not
specified, and all dependencies support the method 'auto', which is the
same as not specifying a method.

The dependencies which do support multiple detection methods
additionally support other values, depending on the dependency.

The Qt module's preprocess function also uses a dependency to fix preprocessing tools, so it also now takes a method argument, which is passed on to the dependency finder.

Closes #1472 
Closes #711

### CODE REVIEW

File: mesonbuild/dependencies.py [hunk 1]
```diff
@@ -1490,6 +1534,14 @@ def get_compile_args(self):
     def get_link_args(self):
         return self.libs
 
+    def get_methods(self):
+        if mesonlib.is_windows():
+            return ['pkgconfig', 'sysconfig']
+        elif mesonlib.is_osx():
+            return ['pkgconfig', 'extraframework']
+        else:
+            return ['pkgconfig']
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies.py]: I would call it `pkg-config` since `pkgconfig` is just one of implementations

**Comment by ignatenkobrain** [comment]
I think it would be better to use enum instead of plain strings

File: mesonbuild/dependencies.py (hunk 1 above)
Comment (absmall) [inline mesonbuild/dependencies.py]: Done

**Comment by absmall** [comment]
I'm not sure of what value switching to enum provides. I can see that it would help to make sure that if two dependency types had a shared method, that they would both have the same name, not spelled a little bit differently, and that could make them easier to remember, but I don't see anything other than that, and that is not really necessary. Having them in an enum will be harder if dependency.py is eventually broken up into separate files, it would make creating a new dependency less dynamic.

I could be missing some advantage, but it seems to me like the drawbacks are more than the advantages.

**Comment by nirbheek** [comment]
The main advantage is that it self-documents all the dependency methods in one place.

**Comment by absmall** [comment]
I'm not really convinced, I don't see the value of documenting all possible methods when each dependency type supports only a subset of them, and what subset that is for each dependency isn't documented. But okay, I'll make that change when I have a chance.

**Comment by jpakkane** [comment]
Enums are great for languages that have native support for them. Sadly Python is not one of these languages. :(

**Comment by nirbheek** [comment]
Python does have [native support for enums](https://docs.python.org/3/library/enum.html), what do you mean? I even [used it for tests](https://github.com/mesonbuild/meson/pull/1575/files#diff-0e150adfe1f55c100b4edc3a4b107418R27).

**Comment by jpakkane** [comment]
I meant more like having a switch statement and enforcing that all values are handled and all that. (enum in the language proper, not enum as a library addition)

**Comment by nirbheek** [comment]
Ah that way, yes. I believe CodeCI will warn if all enum types aren't handled in an if/else ladder though. If not, it should. ;)

**Comment by jpakkane** [comment]
At the beginning of `find_external_dependency` check if `method` is defined and if it is, that its value is a string. Something like:

    if not isinstance(kwargs.get('method', 'dummy'), str):
        # raise exception here

File: mesonbuild/dependencies.py [hunk 2]
```diff
@@ -33,11 +34,35 @@
...
+        method = DependencyMethods(kwargs.get('method', 'auto'))
+
+        # Set the detection method. If the method is set to auto, use any available method.
+        # If method is set to a specific string, allow only that detection method.
+        if method == DependencyMethods.AUTO:
+            self.methods = self.get_methods()
+        elif method in self.get_methods():
+            self.methods = [method]
+        else:
+            raise MesonException('Unsupported detection method: {}, allowed methods are {}'.format(method.value, mlog.format_list(map(lambda  x: x.value, [DependencyMethods.AUTO] + self.get_methods()))))
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies.py]: __[flake8]__
* [E271] multiple spaces after keyword

**Comment by jpakkane** [comment]
This improves the UX nicely. Thanks.

**Pull request merged**